marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/auth/LoginHandler.java

```diff
@@ -45,28 +43,45 @@
     @Override
     public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse httpResponse, Authentication authentication) throws IOException {
         AuthenticationToken token = (AuthenticationToken) authentication;
-
+        Collection<GrantedAuthority> grantedAuthorities = token.getAuthorities();
+        final boolean[] hasLoginAuthority = {false};
         ObjectMapper mapper = new ObjectMapper();
 
         ObjectNode jsonResponse = mapper.createObjectNode();
 
         List<TextNode> authorities = new ArrayList<>();
-        token.getAuthorities().forEach(auth -> {
+
+        grantedAuthorities.forEach(auth -> {
             String authority = auth.getAuthority();
             if (authority.length() > 5) { //trim prefix "ROLE_"
-                authorities.add(new TextNode(auth.getAuthority().substring(5, authority.length())));
+                String authorityName = authority.substring(5);
+                hasLoginAuthority[0] = ("loginToHubCentral".equals(authorityName)) || hasLoginAuthority[0];
+                authorities.add(new TextNode(authorityName));
             }
         });
-        jsonResponse.putArray("authorities").addAll(authorities);
-
-        clearAuthenticationAttributes(request);
 
-        httpResponse.setContentType("application/json");
-        httpResponse.getOutputStream().write(mapper.writeValueAsBytes(jsonResponse));
+        if (!hasLoginAuthority[0]) {
+            sendUnauthorizedResponse(httpResponse);
+        }
```

Comment:
Can we authenticate in the AuthenticationFilter? My understanding of spring security is that all authentications should be performed in the authentication filters.

Since spring performs a lot of stuff under the hood, have we analyzed the consequences of authenticating outside the filter chain in the onAuthenticationSuccess handler?

One thing I found is that spring security publishes an InteractiveAuthenticationSuccessEvent on successful authentication. As we authenticate in the onAuthenticationSuccess handler, from the spring security perspective, the authentication was successful so we might still be publishing this event, which would create issues if ever setup event listeners for this event.

Reply:
We are not authenticating here, only authorizing authenticated users. User who has reached here is already successfully authenticated. We were throwing "InsufficientAuthenticationException" for users who were successfully authenticated (and throwing 403 using onAuthenticationFailure handler) which as per the previous conversation in the PR is not the right exception to be thrown .